pkg/proxmox/goproxmox/api_client.go

```diff
@@ -151,6 +151,17 @@ func (c *APIClient) DeleteVM(ctx context.Context, nodeName string, vmID int64) (
 		return nil, fmt.Errorf("cannot find node with name %s: %w", nodeName, err)
 	}
 
+	cluster, err := c.Cluster(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("cannot get cluster")
+	}
+
+	if vmidFree, err := cluster.CheckID(ctx, int(vmID)); vmidFree {
+		return nil, fmt.Errorf("vm with id %d does not exist", vmID)
+	} else if err != nil {
+		return nil, fmt.Errorf("cannot check if vmid %d is taken: %w", vmID, err)
+	}
+
```

Comment:
```suggestion
	if vmidFree, err := cluster.CheckID(ctx, int(vmID)); vmidFree {
		return nil, VMIDFreeErr
	} else if err != nil {
		return nil, fmt.Errorf("cannot check if vmid %d is taken: %w", vmID, err)
	}
```

Reply:
You would need to go to `reconcileDelete` and check for that error with

 ```
 if err !=nil && errors.Is(err, VMIDFreeErr) {
  return nil
 }
 else if err != nil {
   return errors.Wrap(err, "cannot delete vm")
}

Reply:
The current implementation actually works in my test setup.
The `DeleteVM` function is called from https://github.com/ionos-cloud/cluster-api-provider-proxmox/blob/7b75c862dd90e833c6ff1e6a34a9e1a7e340f39d/internal/service/vmservice/delete.go#L37
If the resulting error matches the string `does not exist` (checked in the `VMNotFound` function), the VM object is finally removed (by removing the finalizer).
I didn't want to reimplement the error handling of the existing implementation and the error `vm with id %d does not exist` perfectly reflects the result of the check. If we throw a specific error here, we should also do the same for the other cases in `DeleteVM`, shouldn't we?

Reply:
https://github.com/ionos-cloud/cluster-api-provider-proxmox/blob/7b75c862dd90e833c6ff1e6a34a9e1a7e340f39d/internal/service/vmservice/delete.go#L38-L44

Here, you would change `if VMNotFound` to `if errors.Is(err, VMIDFreeErr)`
Or you can change the function `VMNotFound`

Reply:
Perhaps it's better to check them both.

```
if VMNotFound(err) || errors.Is(err, VMIDFreeErr) { 
```

as I started to think about it, it might some race condition happens, and you know the Proxmox API is so lazy

Reply:
Ok, I created a new `VMIDFreeErr` and return/check for it now. I hope I understood your comment correctly.

Which race condition are you thinking of?

Reply:
I'm just concerned if some race conditions happen during the deletion of the machine, 
for example, if the VMID returned is free but the VM is not completely deleted, or vice versa,
we have found that those conditions might happen when engaging with the Proxmox API.